src/systems/filecoin_vm/runtime/impl/runtime.go

```diff
@@ -308,6 +309,38 @@ func (rt *VMContext) _checkNumValidateCalls(x int) {
 	}
 }
 
+func (rt *VMContext) ValidateIsStorageMiner(minerAddr addr.Address) bool {
+	codeID, ok := rt.GetActorCodeID(minerAddr)
+	Assert(ok)
+	if !codeID.IsBuiltin() {
+		return false
+	}
+	return (codeID.As_Builtin() == actor.BuiltinActorID_StorageMiner)
+}
+
+func (rt *VMContext) ValidateImmediateCallerDetermineRecipient(entryAddr addr.Address) addr.Address {
+	if rt.ValidateIsStorageMiner(entryAddr) {
+		// Storage miner actor; implied funds recipient is the associated owner address.
+		ownerAddr, workerAddr := rt.GetMinerAccountsAssert(entryAddr)
+		rt.ValidateImmediateCallerInSet([]addr.Address{ownerAddr, workerAddr})
+		return ownerAddr
+	} else {
+		// Account actor (client); funds recipient is just the entry address itself.
+		rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_Account)
+		return entryAddr
+	}
+}
```

Comment:
No! The Runtime should not have such application-specific coupling to the actors. This is easily implemented outside of the runtime and so should not be here.